### PR TITLE
fix: track range bounds to detect write skew during range reads for SSI

### PR DESCRIPTION
**Pull request by arriqaaq**
## Description

This PR addresses the [issue](https://github.com/surrealdb/surrealkv/issues/28) regarding write skews during range reads when using SSI mode for isolation

### CODE REVIEW

**Review by mumoshu — APPROVED** [review]
Some nits in naming and testing only. LGTM overall. Amazing work @arriqaaq!!

File: src/storage/kv/transaction.rs [hunk 1]
```diff
@@ -1663,4 +1684,102 @@ mod tests {
...
+        let key4 = Bytes::from("k4");
+        let key5 = Bytes::from("k5");
+        let key6 = Bytes::from("k6");
+        let key7 = Bytes::from("k7");
+        let value1 = Bytes::from("v1");
+        let value2 = Bytes::from("v2");
+        let value3 = Bytes::from("v3");
+        let value4 = Bytes::from("v4");
+
+        // concurrent inserts between scans should fail
```
[the remark this answers is not in the document]
Comment (arriqaaq) [reply src/storage/kv/transaction.rs:1701]: perfectly makes sense!

File: src/storage/kv/transaction.rs [hunk 2]
```diff
@@ -1663,4 +1684,102 @@ mod tests {
             txn3.commit().await.unwrap();
         }
     }
+
+    async fn g2_item_predicate(is_ssi: bool) {
```
[the remark this answers is not in the document]
Comment (arriqaaq) [reply src/storage/kv/transaction.rs:1688]: Yup! I derived based on the error you pointed out just to keep the naming convention similar to the hermitage tests!

File: src/storage/kv/transaction.rs [hunk 3]
```diff
@@ -88,6 +88,8 @@ pub struct Transaction {
     /// `read_set` is the keys that are read in the transaction from the snapshot. This is used for conflict detection.
     pub(crate) read_set: Mutex<Vec<(Bytes, u64)>>,
 
+    pub(crate) range_bouds: Mutex<Vec<(Bound<VariableSizeKey>, Bound<VariableSizeKey>)>>,
```
[the remark this answers is not in the document]
Comment (arriqaaq) [reply src/storage/kv/transaction.rs:91]: Totally makes sense, thank you for the suggestion

File: src/storage/kv/transaction.rs [hunk 4]
```diff
@@ -1663,4 +1684,102 @@ mod tests {
...
+            assert_eq!(res.len(), 2);
+            assert_eq!(res[0].1, value1);
+            assert_eq!(res[1].1, value2);
+
+            txn1.set(&key3, &value3).unwrap();
+            txn2.set(&key4, &value4).unwrap();
+
+            txn1.commit().await.unwrap();
+
+            assert!(match txn2.commit().await {
```
[the remark this answers is not in the document]
Comment (arriqaaq) [reply src/storage/kv/transaction.rs:1714]: I think this is a very good refactor! Thanks for the detailed suggestion. I am thinking it is bit explicit to read the test and understand in the current way, hence was thinking of keeping it as it is..

**Pull request merged**